### ISSUE TITLE
Tool-tips for the URI of a location in the viewer are not complete URIs

### ISSUE THREAD

**Issue by GabeDeBacker** [microsoft/sarif-vscode-extension#294]
In the viewer, the "result info" tab shows a location that is just a file-name (say "dirs" as the file name).

The tooltip is just showing file://dirs instead of the complete path.

The tooltip should be the complete path.

**Comment by mukul-ms** [comment]
Hey @GabeDeBacker 

I think the reason its just showing `file://dirs` is because that's actually the complete path provided in the sarif file. 

E.g. This is the snippet from the sarif file that issue came from 
- ![image](https://user-images.githubusercontent.com/43248455/87206496-db3a2700-c2d7-11ea-8880-9a939ae53260.png)


I tried it on another test file and it does show the complete path as given in the sarif file 
- ![image](https://user-images.githubusercontent.com/43248455/87206435-b645b400-c2d7-11ea-9038-5f4376299a67.png)

So from our understanding this is the expected output based on the input. Please let me know if a different outcome is expected

cc @jeffersonking

**Comment by GabeDeBacker** [comment]
Interesting.. I was trying to read through the SARIF spec, and I couldn't determine what the "rules" are for the URIs as far as locating artifacts. From the snippets you presented in the issue, it isn't clear to me if there is anything in the SARIF log that would point to the "complete" correct path.

So, it appears to me there is both a bug in the tool that generated this SARIF log and an issue that needs to be addressed in the SARIF spec.

@michaelcfanning - Can you help clarify this?

Things to keep in mind is that the SARIF output is not written next to the source files (the SARIF is placed in a completely different location), so the viewer would not be able to "assume" that the complete path is relative to the location of the SARIF log.

**Comment by michaelcfanning** [comment]
So what we appear to have here are relative paths from some unknown root. Tool providers can help by adding a uriBaseId with the relative path the helps establish which relative URIs are associated with a common base. It looks like that isn't happening here. If the tool is emitting relative paths in the log file that are relative to different base URLs, that's a problem the tool needs to resolve.

If all these relative paths are implicitly based on a common root, the viewer should query the user for that root and then remap all the log file URIs.

@lgolding, who handles spec concerns.

**Comment by jeffersonking** [comment]
From a viewer standpoint, this is "by design".  We can only show as much as the log tells us (I looked at the full original log). 

WRT "viewer should query the user": current design defers that until (if) the file contents are opened. The alternative (as it previously was) was considered distracting and blocking.

**Comment by lgolding** [comment]
There is a lot to say here :-)

First: `artifactLocation.uri` is allowed to be either a relative reference (for example, `some_dir/some_file.c`) or an absolute URI (for example, `file:///C:/dev/my-project/some_dir/some_file.c`).

If `artifactLocation.uri` is an absolute URI, you are done. The viewer can do whatever it wants (for example, display the full URI, or display just the file name part and provide the full URI as a tooltip, or anything else).

If `artifactLocation.uri` is a relative reference, there are a couple of options:

1. If `artificatLocation.uriBaseId` is present, its value is a symbolic name (such as `SOURCE_ROOT`) that stands for the absolute path _with respect to which_ the relative reference was expressed. And now there are two possibilities:

    a. The property `run.originalUriBaseIds` tells you what `SOURCE_ROOT` stood for _on the machine where the scan was performed_. A viewer is allowed to see if that path works on the current machine as well. If it does, great! If not, the viewer can prompt to find out what `SOURCE_ROOT` means on this machine.

    b. `run.originalUriBaseIds` is absent, or it is present but it doesn't mention `SOURCE_ROOT`. The viewer should prompt to find out where `SOURCE_ROOT` is.

2. If `artifactLocation.uriBaseId` is absent, the viewer can prompt to find out what base path it should use for `artifactLocation`s that _don't specify_ a `uriBaseId`.

After all that, the viewer has either figured out the absolute path, or it hasn't. If it has, then great! It has all the display options I mentioned above. If it hasn't, it should display what it has. It could display the full "relative path" (`some_dir\some_file.c`). It could display just the file name `some_file.c`, and display the full "relative path" as a tooltip. Or whatever.

What the viewer _cannot_ do -- and that's why I reopened the bug -- is just prepend `file:///` to the relative path and call it good. If the scheme is present, you must supply the absolute path. If you don't have the absolute path, you can't supply the scheme.

Hope that helps.

**Comment by jeffersonking** [comment]
Note: The viewer does account for (1) and (2) as mentioned by @lgolding. However, that information was not present in the SARIF log (not shared for privacy reasons). We don't prompt immediately per the design considerations in my previous comment.

Regarding the pre-pending of `file:///` to relative paths:
* @GabeDeBacker is this the spirit of this bug? It seemed you were expecting more path segments.
* Implementation detail: We use VS Code's URI object and it does not round-trip relative URIs. It didn't seem cost effective to re-implement URI (or search for a compatible library) for such cases. @lgolding You're welcome to push back here.
* Re "What the viewer *cannot* do...", I'm going to suggest *it can*. Going full-lawyer here: §3.4.4 says "If `uriBaseId` is not yet resolved, the consumer MAY use other information or **heuristics** to locate the artifact." I would argue prepending `file:///` is a fair (albeit overly-simple) heuristic, and thus within the spec.

**Comment by lgolding** [comment]
But you are _not_ "[using] other information or heuristics to locate the artifact". You are just displaying a string. Suppose the viewer were to say to itself, "Hmmm... I wonder if I could find this file by prepending the `file:` scheme. Let's find out..." ... and then suppose that turned out to _work_. Then sure, the viewer would be entitled to conclude that `file:///whatever-you-had-before` really _was_ the location, and to display it wherever it wanted. But that's not what you did.

**Comment by jeffersonking** [comment]
@lgolding: That's fair. :-) The first two bullet points are the more practical concerns though.

**Comment by GabeDeBacker** [comment]
@lgolding  - All I'm saying is the SARIF spec to say something like "all artifact locations MUST resolve to a fully qualified URI using the information present in the SARIF log.

Without this statement, tools can output files that cannot be remapped from a "remote path" (say from a CI build system path) to a "local path" present on a developers machine. Which makes "path remapping" by the SARIF viewers difficult to impement.

For example, there are 69k files named "dirs" in our repository.

**Comment by lgolding** [comment]
@GabeDeBacker I agree that there's something wrong if the file contains results from two distinct files, both named `dirs`, that live in different directories, and the file provides nothing but the file names. The file should at least provide distinct relative paths:

```
"results": [
  {
    "locations": [
      "physicalLocation": {
        "artifactLocation": {
          "uri": "some/folder/dirs"
        }
      }
    ]
  },
  {
    "locations": [
      "physicalLocation": {
        "artifactLocation": {
          "uri": "other/directory/dirs"
        }
      }
    ]
  }
]
```

Feel free to file a suggestion in the SARIF TC's GitHub repo: https://github.com/oasis-tcs/sarif-spec. (I think anyone can do that.)

Thanks for the idea,
Larry
